infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,281 @@
+            event.getSourceLeadershipManager().notifyAll();
+        }
+    }
+
+    private void processRuntimeTask(LogReplicationRuntimeEvent event) {
+        LogReplicationSession session = event.getRuntimeFsm().getSession();
+        // for a given session, The fsm event should be processed in the order they are queued. This condition ensures
+        // that in the event of 2 threads contending for the monitor, only the task submitted first would be processed.
+        synchronized (event.getRuntimeFsm()) {
+            while (!sessionToRuntimeEventIdMap.get(session).get(0).equals(event.getEventId())) {
+                try {
+                    event.getRuntimeFsm().wait();
+                } catch (InterruptedException e) {
+                    log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+                }
+            }
+        }
+
+        LogReplicationRuntimeState currState = event.getRuntimeFsm().getState();
+        if (currState.getType() == LogReplicationRuntimeStateType.STOPPED) {
+            log.info("Log Replication Communication State Machine has been stopped. No more events will be processed.");
+            return;
+        }
+
+
+        try {
+            LogReplicationRuntimeState newState = currState.processEvent(event);
+            if (newState != null) {
+                event.getRuntimeFsm().transition(currState, newState);
+                event.getRuntimeFsm().setState(newState);
+
+            }
+        } catch (IllegalTransitionException illegalState) {
+            log.error("Illegal log replication event {} when in state {}", event.getType(), currState.getType());
+        }
+
+        synchronized(event.getRuntimeFsm()) {
+            sessionToRuntimeEventIdMap.get(session).remove(0);
+            event.getRuntimeFsm().notifyAll();
+        }
+    }
+
+    private boolean canProcessEventForSession(LogReplicationSession session, UUID currEventId) {
+            return (sessionToReplicationEventIdMap.get(session) != null &&
+                        !sessionToReplicationEventIdMap.get(session).isEmpty() &&
+                        sessionToReplicationEventIdMap.get(session).get(0).equals(currEventId)) ||
+                    (sessionToDelayedReplicationEventIdMap.get(session) != null &&
+                            !sessionToDelayedReplicationEventIdMap.get(session).isEmpty() &&
+                            sessionToDelayedReplicationEventIdMap.get(session).get(0).equals(currEventId));
+    }
+
+    private void processReplicationTask(LogReplicationEvent currEvent) {
+        LogReplicationSession session = currEvent.getReplicationFsm().getSession();
+        // for a given session, the fsm events should be processed in the order they are queued. This snippets ensures
+        // that in the event of 2 threads contending for the monitor, only the task submitted first would be processed.
+        synchronized(currEvent.getReplicationFsm()) {
+            while (true) {
```

Comment:
We need the loop because say thread2 and thread3 both are waiting. Thread1, after processing, notifies both the waiting threads.
Now only one of them should actually run while the other goes back to waiting. So we need to to check which one is in position 1 in the order-list.

Reply:
ok, I see.  I got confused because the approach in waiting for the runtime and sink tasks is implemented in a different way.  Both do the same thing though.  If you make further changes to the PR, I suggest keeping it the same as the other 2.